pkg/controller/mcmhub/hook.go

```diff
@@ -0,0 +1,289 @@
+
+	if len(preHook) == 0 && len(postHook) == 0 {
+		return nil
+	}
+
+	//check if the subIns have being changed compare to the hook registry
+	if !a.isUpdateSubscription(subIns) {
+		return nil
+	}
+
+	jobs, err := a.gitClt.DownloadHookResource(subIns)
+
+	if err != nil {
+		return err
+	}
+
+	printJobs(jobs, a.logger)
+	//update the base Ansible job and append a generated job to the preHooks
+	return a.addHookToResitry(subIns, jobs)
+}
+
+func printJobs(jobs []unstructured.Unstructured, logger logr.Logger) {
+	for _, job := range jobs {
+		logger.V(3).Info(fmt.Sprintf("download jobs %#v", job))
+	}
+}
+
+type SuffixFunc func(*subv1.Subscription) string
+
+func suffixFromUUID(subIns *subv1.Subscription) string {
+	return fmt.Sprintf("-%s", subIns.GetUID()[:5])
+}
+
+func (a *AnsibleHooks) addHookToResitry(subIns *subv1.Subscription, jobs []unstructured.Unstructured) error {
+	a.logger.V(2).Info("entry addNewHook subscription")
+
+	if len(jobs) == 0 {
+		return fmt.Errorf("failed to get the prehook from git")
+	}
+
+	defer a.logger.V(2).Info("exit addNewHook subscription")
+
+	subKey := types.NamespacedName{Name: subIns.GetName(), Namespace: subIns.GetNamespace()}
+	preHook := subIns.Spec.Prehook
+	postHook := subIns.Spec.Posthook
+
+	a.registry[subKey] = &Hooks{
+		pre:       unstructured.Unstructured{},
+		post:      unstructured.Unstructured{},
+		lastSub:   subIns,
+		preHooks:  []unstructured.Unstructured{},
+		postHooks: []unstructured.Unstructured{},
+	}
+
+	suffix := a.suffixFunc(subIns)
+	for _, job := range jobs {
+		jobKey := types.NamespacedName{Name: job.GetName(), Namespace: job.GetNamespace()}
+		//need to skip the status, otherwise the creation will fail
+		if strings.EqualFold(jobKey.String(), preHook) {
+			a.registry[subKey].pre = job
+			ins := a.registry[subKey].pre.DeepCopy()
+			ins.SetName(fmt.Sprintf("%s%s", ins.GetName(), suffix))
+			a.registry[subKey].preHooks = append(a.registry[subKey].preHooks, *ins)
+		}
+
+		if strings.EqualFold(jobKey.String(), postHook) {
+			a.registry[subKey].post = job
+			ins := a.registry[subKey].post.DeepCopy()
+			ins.SetName(fmt.Sprintf("%s%s", ins.GetName(), suffix))
+			a.registry[subKey].postHooks = append(a.registry[subKey].postHooks, *ins)
+		}
+	}
+
+	return nil
+}
+
+func (a *AnsibleHooks) ApplyPreHook(subKey types.NamespacedName) (types.NamespacedName, error) {
+	a.logger.WithName(subKey.String()).V(2).Info("entry ApplyPreHook")
+	defer a.logger.WithName(subKey.String()).V(2).Info("exit ApplyPreHook")
+	hks, ok := a.registry[subKey]
+	if ok && len(hks.preHooks) != 0 {
+		t := &hks.preHooks[len(hks.preHooks)-1]
+
+		if err := a.clt.Create(context.TODO(), t); err != nil {
+			if !k8serrors.IsAlreadyExists(err) {
+				return types.NamespacedName{}, err
+			}
+		}
+
+		return types.NamespacedName{Name: t.GetName(), Namespace: t.GetNamespace()}, nil
+	}
+
+	return types.NamespacedName{}, nil
+}
+
+func (a *AnsibleHooks) isUpdateSubscription(subIns *subv1.Subscription) bool {
```

Comment:
The pre-/post- ansible job could update in git repo, the function should allow the new ansibe job to register to the subscription.

Reply:
will add this check later, the check will be based on the commit id of git